pool/pool.go

```diff
@@ -0,0 +1,147 @@
+package pool
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+)
+
+var log = logging.Logger("fula/pool")
+
+type (
+	Pool struct {
+		ctx    context.Context
+		cancel context.CancelFunc
+		wg     sync.WaitGroup
+
+		*options
+		sub   *pubsub.Subscription
+		topic *pubsub.Topic
+	}
+)
+
+func New(o ...Option) (*Pool, error) {
+	opts, err := newOptions(o...)
+	if err != nil {
+		return nil, err
+	}
+	return &Pool{
+		options: opts,
+	}, nil
+}
+
+func (p *Pool) Start(ctx context.Context) error {
+	gsub, err := pubsub.NewGossipSub(ctx, p.h,
+		pubsub.WithPeerExchange(true),
+		pubsub.WithFloodPublish(true),
+	)
+	if err != nil {
+		return err
+	}
+	// TODO: implement topic validators once there is some chain integration.
```

Comment:
I didn't catch what "topic valdiator" validates

Reply:
gosspisub offers a way to [validate](https://github.com/libp2p/go-libp2p-pubsub/blob/master/validation.go) weather a message is from legitimate nodes. This is useful to avoid propagating messages published by non-participants. Remember; gossipsub with flood publish just forwards the messages sent so we need some way to differentiate what to forward and what to not forward.

An example usage of validators can be found in [Lotus](https://github.com/filecoin-project/lotus/blob/0770bced090aef6e9feaf33fc42198984df489d2/node/modules/services.go#L158), where things like topic and miner ID are validated against the chain to confirm that the sender of a message is indeed a lotus node.